### ISSUE TITLE
Question about neighbor list pair fetching from CUDA platform

### ISSUE THREAD

**Issue by dingye18** [openmm/openmm#4067]
Hello for OpenMM developers!

I'm review the code in findInteractingBlocks.cu for neighbor list information fetching in these days.
I have two questions for interacting blocks searching.

1. Refer to the literature [Efficient Nonbonded Interactions for Molecular Dynamics on a Graphics Processing Unit](https://doi.org/10.1002/jcc.21413). The atoms within the system are reordered to optimize coherence. Protein and water molecules are separately processed in practice. Water molecules are reordered periodically in simulation to ensure spatial coherency. However, I didn't found any code that is related with this optimization in OpenMM. Does this optimization is still working in OpenMM interacting blocks construction?
2.  Refer to the code in nonbonded.cu [Line 297](https://github.com/openmm/openmm/blame/ed049af1b34cab4dd47f20dcfddb99bcad6c7913/platforms/cuda/src/kernels/nonbonded.cu#L297) and [332](https://github.com/openmm/openmm/blame/ed049af1b34cab4dd47f20dcfddb99bcad6c7913/platforms/cuda/src/kernels/nonbonded.cu#L332).
I'm wondering that if the following code can be used to reconstruct the neighbor list pair information from interacting blocks and interacting atoms in host device with.
Should I append the value in single pairs array into the `neighborList_cuda` too?

```c++
int num_interacting_tiles = cu.getNonbondedUtilities().getInteractingTiles().getSize();
int num_interacting_atoms = cu.getNonbondedUtilities().getInteractingAtoms().getSize();
int num_interaction_count = cu.getNonbondedUtilities().getInteractionCount().getSize();
vector<int> interaction_count(num_interaction_count);
vector<int> interacting_tiles(num_interacting_tiles);
vector<int> interacting_atoms(num_interacting_atoms);
cu.getNonbondedUtilities().getInteractingTiles().download(interacting_tiles);
cu.getNonbondedUtilities().getInteractingAtoms().download(interacting_atoms);
cu.getNonbondedUtilities().getInteractionCount().download(interaction_count);

OpenMM::NeighborList neighborList_cuda;
for (int pos = 0; pos < interaction_count[0];pos++){
    int x = interacting_tiles[pos];
    for (int ii = 0; ii < 32; ii++){
        int atomI = x * 32 + ii;
        for(int jj = 0; jj < 32; jj++){
            int atomJ = interacting_atoms[pos * 32 + jj];
            neighborList_cuda.push_back(AtomPair(atomI, atomJ));
        }
    }
}
```

**Comment by dingye18** [comment]
I'm developing the [openmm-deepmd-plugin](https://github.com/JingHuangLab/openmm_deepmd_plugin) for Deep Potential model. The current implementation not only support the simulation with DP model by OpenMM, but also the DP/MM simulation with fixed/adaptive changed DP region. I'm trying to optimize the plugin performance with the fancy neighbor list searching algorithm in OpenMM. That's why I'm working on this.

I did noticed that you are working on the NNPOps project also. However, since the Deep Potential model are built on TensorFlow and self-defined operators. The current implementation of NNPOps seems not compatible with Deep Potential model.

**Comment by peastman** [comment]
That sounds interesting!  Be aware that if you rely on the internals of OpenMM's neighbor list, you may need to update your code regularly.  Those internal details aren't considered part of the public interface, so they can change with time.

Reordering is done by `ComputeContext::reorderAtoms()`.  When we come to building the neighbor list, that code doesn't need to know anything about how the atoms were reordered.  It just computes the interactions for whatever ordering it has.  The goal of reordering is to make sure that when `findBlockBounds()` computes bounding boxes for sets of 32 atoms, they'll be as small as possible.

The full neighbor list structure is more complicated than what you show.  I suggest studying `nonbonded.cu` until you're sure you understand it.  There are several cases that get handled separately.

- Diagonal tiles, involving a fixed set of 32 sequential atoms with the same 32 atoms.  Only half the pairs need to be computed to avoid double counting interactions (i with j and j with i).  These tiles have exclusions that need to be considered.
- Other tiles with exclusions.  These involve a fixed set of 32 atoms with a different set of 32 atoms.  All pairs in these tiles get calculated, but again you need to consider exclusions.
- Tiles without exclusions.  These involve a set of 32 sequential atoms with an arbitrary (not necessarily sequential) set of 32 other atoms.  These are the ones that get identified by `findBlocksWithInteractions()`.
- Single pairs that get calculated individually.

Also consider whether it's worth the trouble of using the complicated neighbor list structure, or whether you can just write your own much simpler one.  In classical force fields, each interaction is very cheap.  Building the neighbor list can represent a significant fraction of the time, so we try really hard to optimize it.  With NNPs, calculating the interactions is far more expensive.  Any reasonable algorithm for building the neighbor list will take negligible time in comparison.

**Comment by dingye18** [comment]
Thanks for your response and comments. Since DeePMD-kit provide the api for MD engine to call with neighbor list in input and make some improvement with LAMMPS ( 4.6 ns/day for 64 water molecules ). I do really hope to optimize DP+OpenMM (0.985 ns/day for 256 water molecules without neighborlist) with neighbor list so that it's comparable with DP+LAMMPS in simulation performance.

**Comment by peastman** [comment]
How do the speeds compare when simulating the same number of molecules?

I suggest profiling to see how the time is spent.  How long does building the neighbor list take?  How long does computing the interaction take once you have the neighbor list?  That tells you what the potential benefit is from a faster neighbor list builder.

**Comment by dingye18** [comment]
You’re right, I’ve been running some tests on a single A40 GPU to compare the performance of the OpenMM DP Plugin and LAMMPS+DP recently. The results for LAMMPS+DP showed that building the neighbor list accounted for 9-15% of the simulation time. Meanwhile, OpenMM+DP had a comparable simulation performance to LAMMPS+DP, but it required more GPU memory for the neighbor list construction in DP operators. As a result, simulations with a large number of atoms may not be feasible with OpenMM+DP due to memory limitations.

| Number of Atoms | 6144 | 12288 | 18432 | 24576 |
|--------|--------|--------|--------|--------|
| OMM+DP (ns/day) | 0.243 | 0.121 | 0.074 | 0.056 |
| LAMMPS+DP (ns/day) | 0.228 | 0.123 | 0.084 | 0.063 |
| Fraction of Neigh. List Cost in LAMMPS| 13.27% | 10.68% | 9.77% | 9.32% |

**Comment by peastman** [comment]
In standard MD simulations, 24,000 atoms is considered quite small.  OpenMM has no trouble simulating systems 100 times larger, even on consumer GPUs with limited memory.  If you're running into memory limits building the neighbor list, that suggests you're doing some part of it in an inefficient way.

**Comment by dingye18** [comment]
I do agree with you. It’s widely recognized that neural network potentials (NNPs) require significantly more GPU memory than classical force fields in molecular dynamics simulations. As a result, the number of atoms that can be simulated with OpenMM and NNPs is typically much smaller than with classical force fields. This issue is particularly pronounced in the case of the OpenMM+DP plugin, where the inefficient implementation of the neighbor list building algorithm exacerbates the problem in terms of memory usage, although the simulation speed is not highly impacted.

There are still advantages to extracting the neighbor list construction algorithm in OpenMM as an independent module, particularly for simulations involving NNPs.

**Comment by peastman** [comment]
The point is that the neighbor list should take negligible memory.  If that isn't the case, it means there's a problem in your code that you should fix.  It isn't a problem with the OpenMM neighbor list code, which can handle millions of atoms.
